### PR TITLE
Non-record: Fused Softcap+CE Megakernel (1.94x vs torch.compile) + N-gram Backoff

### PR DESCRIPTION
**Pull request by anthony-maio**
## Non-Record Submission: Makora-Generated Fused CUDA Kernel

Answers OpenAI's explicit request for **megakernels**.

### Fused Softcap + Cross Entropy CUDA Kernel
Generated via [Makora](https://makora.com) automated kernel generation. Fuses `30*tanh(logits/30)` + `cross_entropy` into a single CUDA launch:
- **1.94x faster** than torch.compile
- **7.51x faster** than eager PyTorch
- Warp-level online softmax reduction
- bf16 input, float32 accumulation
- Numerically correct to 5 decimal places (max diff: 0.00001431)

Compiled via `torch.utils.cpp_extension.load_inline` at startup — zero external dependencies.

### Architecture
Same 11L VRL+LeakyReLU² stack as PR #889 (0.9642 bpb) with the fused kernel integrated into the sliding window eval path. Training uses standard PyTorch (kernel is forward-only).

### Validation
- Kernel compiled and verified on H100
- `fused_softcap_ce:True` confirmed in training logs
- Correctness: max absolute diff vs reference = 0.00001431
- 8xH100 full run pending (GPUs currently unavailable)

### Why This Matters
Custom CUDA kernels at this model scale typically lose to torch.compile (we proved this ourselves with 8 Makora kernels on Day 1 — all added overhead). The softcap+CE fusion is the exception because it eliminates a large intermediate tensor (B*T × V float32 capped logits) that torch.compile cannot optimize away.

### Credits
- Kernel generation: [Makora](https://makora.com)
- N-gram backoff: PR #727 by @Asukabot0
- Base model: PR #414 by @signalrush

### CODE REVIEW

**Comment by MatoTeziTanka** [comment]
### Community Review — Fused Softcap+CE Megakernel + N-gram Backoff

**BPB:** 0.9642 post-ngram (3-seed mean, std 0.0002) / 1.1225 pure-neural stride-64 | **Seeds:** 3 | **Artifact:** 15,981,848 B (seed 1337) | **Track:** non-record | **Compliance:** N-gram PASS, eval-budget FLAG

**What this does:** Takes the author's PR #175-style 11L / 512d / VRL / LeakyReLU(0.5)^2 / GQA / SP1024 stack, and stacks two independently-motivated pieces on top: (1) a Makora-generated fused `softcap + cross_entropy` CUDA kernel (compiled at startup via `torch.utils.cpp_extension.load_inline`, with a pure-PyTorch fallback at L67-71); (2) a multi-order (2-7gram) backoff cache applied after sliding-window eval, with an entropy-adaptive mixing weight `alpha = 0.05 + 0.55 * sigmoid(2*(H - 4))`. Filed correctly as non-record.

**What I found in the code (`records/track_10min_16mb/2026-03-26_NgramBackoff_VRL_LeakyReLU2/train_gpt.py`, SHA `15a5cb8c`):**

- **Fused kernel (L23-71):** bf16 input, fp32 accumulation, warp-level online softmax reduction via `__shfl_xor_sync`, one block per 4 rows. Computes `logf(se)+mx-tc` where `tc` is the capped logit at the target. Numerically consistent with the `30*tanh(x/30)` + `F.cross_entropy` fallback. The try/except at L23-71 is a clean "compile or fall back" pattern — works on CPU (gauntlet below verified the fallback). Used only on the forward/eval path (L961 calls it inside `eval_val`, training still uses the standard loss). Per Issue #1409, custom/fused kernels are explicitly permitted.
- **N-gram cache (`NgramBackoffCache`, L988-1031):** 7 orders, 4M-bucket hash tables per order, XOR-with-primes hashing, `min_count=2`, raw count ratios (no smoothing).
- **Strict prefix on context (L1001-1005):** `_hash_ctx` reads `tokens[pos - ctx_w + k]` for `k in [0, ctx_w)`. With `pos = ws + t + 1` (the absolute token index of the *target*, L1088) and `ctx_w >= 1`, the hashed indices are `[pos-ctx_w, ..., pos-1]` — strictly before the target. No oracle / hindsight lookup.
- **`predict(tokens, pos, target)` (L1019-1032):** `target` is only used to look up `full_count` in the context-conditional hash table — it computes P(target | context), not argmax over targets. This is standard n-gram prediction, not oracle selection. Per Issue #677 (valerio-oai, 2026-03-27), eval caches must be backward-looking with no oracle selection — this code is.
- **Score-before-update at window granularity (L1087-1112):** The scoring loop finishes for a whole window, and only then does `cache.update(all_tokens, scored_up_to, new_end)` (L1111) add the window's tokens to the tables; `scored_up_to` starts at the first assigned window's left edge. So during any token's scoring the cache only contains tokens from strictly-earlier windows. No token ever contributes to its own prediction. Legal under Issue #402 / #677.
- **Mixing (L1094-1098):** `mixed_p = (1 - alpha)*model_p + alpha*ng_p`, floored at 1e-12, `-log` taken for NLL. Linear in probability space, standard.

**Why is the BPB 0.9642 and not ~1.08?**

The README is transparent about this, but it's worth stating plainly for the leaderboard context: 0.9642 is **not** comparable to the sliding-window stride-64 numbers on the main leaderboard. From `train_seed1337.log`:

```
final_int6_sliding_window val_loss:1.8953 val_bpb:1.1225 stride:64 eval_time:102169ms
final_ngram        val_loss:1.6277 val_bpb:0.9640 ngram_eval_time:895349ms
```

The pure-neural stride-64 number is **1.1225**, which is exactly in the range expected for an SP1024 11L VRL stack at 6,765 training steps. The 0.9642 is a *post-processing* number that layers a 2-7gram backoff cache over the neural eval. So the "gap to SOTA" isn't a gap at all — the neural model is at ~1.12 like the rest of the SP1024 pack; the cache buys ~0.16 BPB on top for free (in params/bytes), but at a big time cost.

**Main flag — eval-budget compliance:**

- `ngram_eval_time: 895,349 ms` (seed 1337) ≈ **14.9 minutes** on 8×H100 SXM. The stride-64 neural eval alone is already 102,169 ms (~1.7 min). Per the README and Issue #17, eval wallclock on 8×H100 SXM is 10 min.
- The submission README (at `records/track_10min_16mb/.../README.md`, "Compliance" section) frames this as "~15 min on 8×H100 SXM (under 10 min per-GPU)". That "per-GPU" reading is not how I read the rules — the budget is total wallclock on the 8-GPU eval image, not per-device. **I'd want a mod ruling on this before treating 0.9642 as a comparable number, even for non-record.**
- This is the single biggest caveat on the submission. Everything else checks out.

**Gauntlet (CPU pre-flight on the PR head):**
```
[PASS] Import, Hyperparameters (dim=512, layers=11, heads=8, vocab=1024)
[PASS] Model: 26,993,766 params
[PASS] Forward pass: loss=6.9368
[PASS] Artifact: 4,649,574 B (29.1% of 16MB) via int6+lzma on freshly-initialized weights
[INFO] Code size (as downloaded from GitHub raw): 70,446 B
       Submission.json reports bytes_code=67,048 B — 3.4KB delta is GitHub raw encoding vs author's local encode; author's own code-size measurement is what counts per README
[INFO] Est. 8×H100: 45.9 ms/step, 13,058 steps in 10 min
```
Gauntlet PASS; the kernel's `except Exception: fallback` path (L67-71) is what made CPU execution possible — clean engineering.

**Seed coverage / artifact sizes (per README table, verified against `train_seed1337.log`):**
- seed 1337: 15,981,848 B, post-ngram 0.9640
- seed 42:   15,904,632 B, post-ngram 0.9641
- seed 2025: 15,974,308 B, post-ngram 0.9644
- All under 16,000,000 B. Mean 0.9642, std 0.0002. Pre-ngram stride-64 is 1.1225 / 1.1224 / 1.1231 respectively — tight.

**Questions / flags:**
1. **Eval wallclock.** 895s ngram stage + 102s stride-64 stage is ~16.6 min total on 8×H100 SXM. The author's "10 min per-GPU" reading is non-standard. **I'd want @0hq / @valerio-oai to weigh in on whether non-record submissions are bound by the 10-min eval wallclock.** If they are, the ngram stage as written needs to be either sped up (Numba/C++/CUDA the inner `predict`/`update` — right now it's pure Python over a NumPy buffer) or dropped.
2. **Prior-art credit.** README credits PR #727 (@Asukabot0) for the n-gram backoff approach, PR #414 for the neural base, PR #493/#518 for LeakyReLU², and PR #287 for XSA — good attribution discipline, no concerns.
3. **Makora disclosure.** The kernel is explicitly labeled as Makora-generated with a link, which is the right level of disclosure. No issue.

**Verdict:** **NEEDS CLARIFICATION** — on eval-budget interpretation. The technique is compliant (n-gram is backward-looking, score-first, no oracle; kernel is a legal fused op per #1409) and the engineering is clean, but the non-record filing only makes sense if ~15-minute evals are acceptable on this track.

**Recommendation to @cocohearts @valerio-oai @0hq @yuzhougu-oai @notapplica:**
- **MERGE (non-record)** on the strength of the megakernel contribution alone (this was explicitly README-requested; it's a clean 1.94× vs torch.compile result, correctness verified on H100, CPU fallback is drop-in). The fused kernel is worth merging independent of the n-gram result.
- **HOLD** on treating 0.9642 as a comparable n-gram BPB until a mod clarifies the 10-min eval budget for non-record submissions with post-processing caches (Issue #17 / README). A one-line ruling would unblock this and ~30 other ngram-cache PRs in the queue.
- If the budget is 10 min wallclock, I'd ask the author to either (a) profile/port the n-gram inner loops out of Python, or (b) re-file the fused kernel as its own standalone non-record and drop the ngram stage from this PR.

---
*Reviewed by [@MatoTeziTanka](https://github.com/MatoTeziTanka) — [The Agora](https://matotezitanka.github.io/parameter-golf). Gauntlet ran clean on CPU: all 10 checks PASS, artifact budget 29.1%, fused-CE fallback path exercised successfully. AI tooling: review drafted with Claude Code (Opus) using an internal review template; all citations, file paths, and compliance audits were verified against the PR's actual code at SHA `15a5cb8c2a5ef20c8498e22dcdec66db7c344330`.*
